### ISSUE TITLE
IMU bias and extrinsics do not converge

### ISSUE THREAD

**Issue by JJQ7** [ethz-asl/rovio#162]
Hi all,
I'm using Ubuntu 16.04 and ROS Kinetic, everything properly worked when i run rovio with the provided test data.
Then I changed some imu related covariance parameters according to the device values provided by the manufacturer and tried to run with my device (IMU: ICM 20602, and a hardware synchronized camera).
Rovio passes the tests of no motion and little motion. However, when test with large motion, the outputs have a large difference with the ground truth. And sometimes, rovio diverges.
I adjust some parameters to make rovio performs better. But it seems that the parameters can not guarantee similar performance with different data sets (recored by the same device).
I observe that the imu bias and imu-camera extrinsics always changes rapidly under different parameters i have tried. I really hope someone could give me some advices. 
Thank you!

**Comment by JJQ7** [comment]
![figure_1](https://user-images.githubusercontent.com/21231214/37552311-7fbbe768-29ed-11e8-87f9-add39226ee9e.png)

**Comment by JJQ7** [comment]
Aditionally, it seems that the feathers are not tracked correctly when the camera shakes.

**Comment by JJQ7** [comment]
#70 
#80 
#81

**Comment by JJQ7** [comment]
I notice something interesting, if I change covariance parameter of attitude prediction, the size of the trajectory will change. And The effect of tracking landmarks seems to have a close relationship with this parameter. However, I still can't get an  ideal result. Could someone give me some advices on these parameters? Thank you !

**Comment by bloesch** [comment]
Something does indeed seem off. In my experience the gyroscope biases typically converge quite reliably (but then again the scale in your plots is quite small). Also a accelerometer bias of 1m/s^2 seems to large. How do the extrinsics look like? Do you use any calibration for extrinsics or do you pass identity?

There is some description on the parameters in the wiki (https://github.com/ethz-asl/rovio/wiki/Configuration). I am not familiar with the IMU you mentioned but in my experience the default noise parameters are typically too low (the IMU employed in the euroc dataset is rather good). So you should probably slightly increase the values (a factor 10-1000 because the unit are squared).

**Comment by JJQ7** [comment]
Test 1
![bias](https://user-images.githubusercontent.com/21231214/38173707-a9c43e2a-35f5-11e8-8d05-f4e1974993be.png)
![extrinsic1](https://user-images.githubusercontent.com/21231214/38173717-ae453ca6-35f5-11e8-980f-54b2e1c3b899.png)
![extrinsic2](https://user-images.githubusercontent.com/21231214/38173718-ae7f13c2-35f5-11e8-86df-17504d304316.png)
![vel](https://user-images.githubusercontent.com/21231214/38173719-aeb9a2d0-35f5-11e8-9915-6e89c064113e.png)

**Comment by JJQ7** [comment]
Test 2 (same device, same parameters)
![bias](https://user-images.githubusercontent.com/21231214/38173976-b5e6ddd6-35f8-11e8-9823-e873d04f7819.png)
![extrinsic2](https://user-images.githubusercontent.com/21231214/38173997-e5a0516a-35f8-11e8-82ff-5222d1a87b12.png)
![extrinsic1](https://user-images.githubusercontent.com/21231214/38173978-bf4da03a-35f8-11e8-8cda-d4c91d848b49.png)

**Comment by JJQ7** [comment]
Thanks for your advice. And sorry for my late reply. My computer is broken so everything is lost...
According to your suggestion, I adjusted the parameters, and conducted two tests (using same device and some parameters. velocity is recorded by Qualisys motion capture system). However, there is still some problems about extrinsic and bias. It seems that rovio is trying to attribute measurement error the the extrinsic and bias.(especially in bias)

**Comment by JJQ7** [comment]
I will try to give a fixed extrinsic provided by Kalibr and upload the results in the future.

**Comment by JJQ7** [comment]
![image](https://user-images.githubusercontent.com/21231214/38298258-7dd2bad6-3829-11e8-9952-a6e0f00cdc3c.png)

**Comment by bloesch** [comment]
It's difficult to say how good the biases/extrinsics should converge. Most of your plots for these values look okay. The steps in the last one (just above) are a weird though. What setup did you use.

Your comment concerning "attributing measurement errors to biases" is valid. The filter decides online what states to correct given the measurement error and this is done based on the estimated uncertainty of all states. So if your bias states (including extrinsics) do not converge well then measurement errors will affected them more. But all of this is very dependent on how you tune the covariances and just taking the covariance specifications from the device is unfortunately rarely enough (due to inaccuracies in the modeling). If you feel the bias states are being affected to strongly you can reduce their prediction noise (at the cost of some robustness w.r.t. bad initial convergence), reduce their initial uncertainty (at the cost of wrong initial guesses), or fix the extrinsics.

**Comment by JJQ7** [comment]
yes! I found even in the pubic data (euroc MH_05), the bias provided by rovio is not a fixed value although the final trajectory is close to the ground truth. This really bothers me for a long time. And some divergences come up in euroc data. Maybe some minor adjustments can improve the performance of rovio.

In the last plot (just above),  the same device was used. The data is record on a quadrotor. The quadrotor moves along the trajectory of a square and stays at the four vertices for a certain period of time. The extrinsics change when quadrotor moves.
